wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointPlugin.java

```diff
@@ -0,0 +1,336 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.plugin.customendpoint;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+import java.util.logging.Logger;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.rds.RdsClient;
+import software.amazon.jdbc.AwsWrapperProperty;
+import software.amazon.jdbc.HostSpec;
+import software.amazon.jdbc.JdbcCallable;
+import software.amazon.jdbc.PluginService;
+import software.amazon.jdbc.PropertyDefinition;
+import software.amazon.jdbc.authentication.AwsCredentialsManager;
+import software.amazon.jdbc.plugin.AbstractConnectionPlugin;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.RdsUtils;
+import software.amazon.jdbc.util.RegionUtils;
+import software.amazon.jdbc.util.SlidingExpirationCacheWithCleanupThread;
+import software.amazon.jdbc.util.StringUtils;
+import software.amazon.jdbc.util.SubscribedMethodHelper;
+import software.amazon.jdbc.util.WrapperUtils;
+import software.amazon.jdbc.util.telemetry.TelemetryCounter;
+import software.amazon.jdbc.util.telemetry.TelemetryFactory;
+
+/**
+ * A plugin that analyzes custom endpoints for custom endpoint information and custom endpoint changes, such as adding
+ * or removing an instance in the custom endpoint.
+ */
+public class CustomEndpointPlugin extends AbstractConnectionPlugin {
+  private static final Logger LOGGER = Logger.getLogger(CustomEndpointPlugin.class.getName());
+  private static final String TELEMETRY_WAIT_FOR_INFO_COUNTER = "customEndpoint.waitForInfo.counter";
+
+  protected static final long CACHE_CLEANUP_RATE_NANO = TimeUnit.MINUTES.toNanos(1);
+  protected static final RegionUtils regionUtils = new RegionUtils();
+  protected static final SlidingExpirationCacheWithCleanupThread<String, CustomEndpointMonitor> monitors =
+      new SlidingExpirationCacheWithCleanupThread<>(
+          CustomEndpointMonitor::shouldDispose,
+          (monitor) -> {
+            try {
+              monitor.close();
+            } catch (Exception ex) {
+              // ignore
+            }
+          },
+          CACHE_CLEANUP_RATE_NANO);
+
+  private static final Set<String> subscribedMethods =
+      Collections.unmodifiableSet(new HashSet<String>() {
+        {
+          addAll(SubscribedMethodHelper.NETWORK_BOUND_METHODS);
+          add("connect");
+          add("forceConnect");
```

Comment:
I don't think we need forceConnect(). It's usually used by various monitoring threads and failover.